src/soc/ibm/power9/istep_8_1.c

```diff
@@ -0,0 +1,337 @@
+		OB1_PLL_BUCKET_START      = 26,
+		OB1_PLL_BUCKET_LEN        = 2,
+		OB2_PLL_BUCKET_START      = 28,
+		OB2_PLL_BUCKET_LEN        = 2,
+		OB3_PLL_BUCKET_START      = 30,
+		OB3_PLL_BUCKET_LEN        = 2,
+
+		/* SCRATCH_REGISTER_3 */
+		BOOT_FLAGS_START          = 0,
+		BOOT_FLAGS_LEN            = 32,
+		RISK_LEVEL_START          = 28,
+		RISK_LEVEL_LEN            = 4,
+
+		/* SCRATCH_REGISTER_4 */
+		BOOT_FREQ_MULT_START      = 0,
+		BOOT_FREQ_MULT_LEN        = 16,
+		CP_FILTER_BYPASS          = 16,
+		SS_FILTER_BYPASS          = 17,
+		IO_FILTER_BYPASS          = 18,
+		DPLL_BYPASS               = 19,
+		NEST_MEM_X_O_PCI_BYPASS   = 20,
+		OBUS_RATIO_VALUE          = 21,
+		NEST_PLL_BUCKET_START     = 29,
+		NEST_PLL_BUCKET_LEN       = 3,
+
+		/* SCRATCH_REGISTER_5 */
+		PLL_MUX_START             = 12,
+		PLL_MUX_LEN               = 20,
+		CC_IPL                    = 0,
+		INIT_ALL_CORES            = 1,
+		RISK_LEVEL_BIT_DEPRECATED = 2,
+		DISABLE_HBBL_VECTORS      = 3,
+		MC_SYNC_MODE              = 4,
+		SLOW_PCI_REF_CLOCK        = 5,
+
+		/* SCRATCH_REGISTER_6 */
+		SMF_CONFIG                     = 16,
+		PROC_EFF_FABRIC_GROUP_ID_START = 17,
+		PROC_EFF_FABRIC_GROUP_ID_LEN   = 3,
+		PROC_EFF_FABRIC_CHIP_ID_START  = 20,
+		PROC_EFF_FABRIC_CHIP_ID_LEN    = 3,
+		PUMP_CHIP_IS_GROUP             = 23,
+		SLAVE_CHIP_SBE                 = 24,
+		PROC_FABRIC_GROUP_ID_START     = 26,
+		PROC_FABRIC_GROUP_ID_LEN       = 3,
+		PROC_FABRIC_CHIP_ID_START      = 29,
+		PROC_FABRIC_CHIP_ID_LEN        = 3,
+		PROC_MEM_TO_USE_START          = 1,
+		PROC_MEM_TO_USE_LEN            = 6,
+	};
+
+	uint64_t scratch;
+
+	uint32_t boot_flags;
+	uint8_t risk_level;
+
+	uint8_t eq_gard;
+	uint32_t ec_gard;
+
+	compute_chip_gards(chip, &eq_gard, &ec_gard);
+
+	/* SCRATCH_REGISTER_1 */
+
+	scratch = PPC_SHIFT(read_cfam(chip, MBOX_SCRATCH_REG1_FSI), 31);
+
+	/* ATTR_EQ_GARD (computed at runtime) */
+	PPC_INSERT(scratch, eq_gard, EQ_GARD_START, EQ_GARD_LEN);
+	/* ATTR_EC_GARD (computed at runtime)*/
+	PPC_INSERT(scratch, ec_gard, EC_GARD_START, EC_GARD_LEN);
+
+	write_cfam(chip, MBOX_SCRATCH_REG1_FSI, scratch >> 32);
+
+	/* SCRATCH_REGISTER_2 */
+
+	scratch = PPC_SHIFT(read_cfam(chip, MBOX_SCRATCH_REG1_FSI + 1), 31);
+
+	/* ATTR_I2C_BUS_DIV_REF (default, talos.xml) */
+	PPC_INSERT(scratch, 0x0003, I2C_BUS_DIV_REF_START, I2C_BUS_DIV_REF_LEN);
+	/* ATTR_MC_PLL_BUCKET (seems to not by relevant for Nimbus) */
+	PPC_INSERT(scratch, 0x00, MC_PLL_BUCKET_START, MC_PLL_BUCKET_LEN);
+
+	/* TODO: set OPTICS_CONFIG_MODE_OBUS[0-3] bits when adding support for OBUS. */
+	/* TODO: set OB[0-3]_PLL_BUCKET bits when adding support for OBUS,
+	 *       see getObusPllBucket() in Hostboot for values of ATTR_OB*_PLL_BUCKET */
+
+	write_cfam(chip, MBOX_SCRATCH_REG1_FSI + 1, scratch >> 32);
+
+	/* SCRATCH_REGISTER_3 */
+
+	scratch = PPC_SHIFT(read_cfam(chip, MBOX_SCRATCH_REG1_FSI + 2), 31);
+
+	boot_flags = (read_scom(MBOX_SCRATCH_REG1 + 2) >> 32);
+	risk_level = (get_dd() < 0x23 ? 0 : 4);
```

Comment:
> maybe that is the reason for stuck CPU frequency

Lack of sensor information looks like a more plausible reason to me. OCC likely needs them to determine CPU load.

> For some reason this is 0 in scratch register for first CPU, even though it is DD23. 

I read somewhere that first CPU is started by BMC, maybe it doesn't know proper risk level. However since Hostboot contains code for upgrading it
```
2339.91027|HWAS_I|I>hwas.C: updateProcCompatibilityRiskLevel: Update RISK_LEVEL from 0 to 4 (EC Level: 0x23, RISK_ORIGIN: MRW, PROC_COMPATIBILITY_REQ: 0)
```
it's probably expected (there is also `NIMBUS_ONLY` in comments).

> Also, this line sets risk level for second CPU based on DD of first CPU, and they may not be the same.

Risk level comes from a *system* attribute, so I guess CPUs should have identical level.

> If it is even possible at this point, only FSI SCOM can be used to read from second CPU.

Haven't tried with that address, but I think FSI SCOM needs CPU to be on. Easy to check though.

Reply:
Ok, so it seems to be right thing to do. I wonder if we should update scratch register for CPU0 or if it is ok to leave it as it is set by BMC.